### PR TITLE
Fix tSNE bugs from numpy 1.13 

### LINKED ISSUE #323: tSNE Value Error when no classes are specified
See [yellowbrick t-SNE fit raises ValueError](https://stackoverflow.com/questions/48950135/yellowbrick-t-sne-fit-raises-valueerror):

The issue has to do with the way numpy arrays are evaluated in 1.13 - we need to change [yellowbrick.text.tsne Line 256](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/text/tsne.py#L256) to `if y is not None and self.classes_ is None`

code to produce:

```python
import pandas as pd
from yellowbrick.text import TSNEVisualizer
from sklearn.datasets import make_classification

## produce random data
X, y = make_classification(n_samples=200, n_features=100,
                       n_informative=20, n_redundant=10,
                       n_classes=3, random_state=42)

## visualize data with t-SNE
tsne = TSNEVisualizer()
tsne.fit(X, y)
tsne.poof()
```

error raised:

```
Traceback (most recent call last):
  File "t.py", line 12, in <module>
    tsne.fit(X, y)
  File "/Users/benjamin/Workspace/ddl/yellowbrick/yellowbrick/text/tsne.py", line 256, in fit
    if y and self.classes_ is None:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

### PR DESCRIPTION
**Pull request by bbengfort**
This work in progress is meant to resolve #323 

So far I've:

- removed ambiguous comparisons to numpy arrays, switching them to `if y is not None` etc. 
- added random_state to the visualizer to engage image comparison tests 

Still to do:

- [x] get the image comparison tests working (for some reason the colors are changing)
- [x] get the stack overflow example producing a visualization 

**UPDATE**:

I've got everything working primarily by using `np.unique` to detect the classes and to remove the feature that allowed `TSNE` to filter classes. Now the SO example works and we have several image comparison tests. 

Documentation has also been updated to reflect the changes.

### CODE REVIEW

File: docs/api/text/tsne.py [hunk 1]
```diff
@@ -25,17 +25,13 @@ def tsne(docs, labels, outpath, **kwargs):
...
     docs   = tfidf.fit_transform(corpus.data)
-    labels = corpus.target
+    target = corpus.target
 
     # Whole corpus visualization
-    tsne(docs, labels, "images/tsne_all_docs.png")
-
-    # Partial corpus visualization
-    # Only visualize the sports, cinema, and gaming classes
-    tsne(docs, labels, "images/tsne_limit_classes.png", classes=['sports', 'cinema', 'gaming'])
```
Comment (bbengfort) [inline docs/api/text/tsne.py]: This file generates figures for the documentation; had to remove the partial corpus visualization, but included an example that uses the `labels` arg to name the documents in the target-less fit.

File: yellowbrick/text/tsne.py [hunk 2]
```diff
@@ -130,43 +132,51 @@ class TSNEVisualizer(TextVisualizer):
...
+        projection faster. Specify ``"svd"`` for sparse data or ``"pca"`` for
+        dense data. If None, the original data set will be used.
 
-    decompose_by : int
+    decompose_by : int, default: 50
         Specify the number of components for preliminary decomposition, by
         default this is 50; the more components, the slower TSNE will be.
 
-    classes : list of strings
+    labels : list of strings
```
Comment (bbengfort) [inline yellowbrick/text/tsne.py]: This is the biggest change; can now specify label names (unless they are detected from `y` on fit), classes are no longer filtered.

File: yellowbrick/text/tsne.py [hunk 3]
```diff
@@ -130,43 +132,51 @@ class TSNEVisualizer(TextVisualizer):
...
         The names of the classes in the target, used to create a legend.
+        Labels must match names of classes in sorted order.
 
     colors : list or tuple of colors
         Specify the colors for each individual class
 
     colormap : string or matplotlib cmap
         Sequential colormap for continuous target
 
+    random_state : int, RandomState instance or None, optional, default: None
```
Comment (bbengfort) [inline yellowbrick/text/tsne.py]: Because TSNE, PCA, and SVD are all stochastic, had to add random_state to get the tests to pass.

File: yellowbrick/text/tsne.py [hunk 4]
```diff
@@ -130,43 +132,51 @@ class TSNEVisualizer(TextVisualizer):
...
         Initialize the TSNE visualizer with visual hyperparameters.
         """
         super(TSNEVisualizer, self).__init__(ax=ax, **kwargs)
 
-        # Visualizer parameters
-        self.classes_ = classes
-        self.n_instances_ = 0
-
         # Visual Parameters
-        # TODO: Only colors currently works to select the colors of classes.
```
Comment (bbengfort) [inline yellowbrick/text/tsne.py]: `resolve_colors` is now fixed, so both colormap and colors works as expected; though this will probably change in the future.

File: yellowbrick/text/tsne.py [hunk 5]
```diff
@@ -319,7 +350,7 @@ def finalize(self, **kwargs):
         self.ax.set_xticks([])
 
         # Add the legend outside of the figure box.
-        if self.classes_:
+        if not all(self.classes_ == np.array([self.NULL_CLASS])):
```
Comment (bbengfort) [inline yellowbrick/text/tsne.py]: This is terrible; basically, a check to see if we should add a legend or not; open to suggestions here.

File: yellowbrick/text/tsne.py [hunk 6]
```diff
@@ -252,13 +267,18 @@ def fit(self, X, y=None, **kwargs):
             Returns the instance of the transformer/visualizer
         """
 
-        # If we don't have classes already stored, store them.
-        if y and self.classes_ is None:
-            self.classes_ = [str(label) for label in set(y)]
+        # Store the classes we observed in y
```
Comment (bbengfort) [inline yellowbrick/text/tsne.py]: Part of the labels, classes_ dance.

**Comment by bbengfort** [comment]
This is ready for review if anyone cares to take a look!

**Comment by ndanielsen** [comment]
@bbengfort tests are passing on this one, what else did you want me to take a look at?

**Comment by bbengfort** [comment]
@ndanielsen I was hoping for a code review and a :shipit: - I put comments throughout the code to point out weird or ugly things if you had any ideas.

**Comment by ndanielsen** [comment]
@bbengfort okay, I'll put on my list of things tomorrow during the coworking session to take a look at

**Comment by lwgray** [comment]
@bbengfort I will give it a once over in the morn unless you really want it done tonight

File: docs/api/text/tsne.py (hunk 1 above)
Comment (lwgray) [reply docs/api/text/tsne.py]: It might be obvious but I suggest you include a docstring that states as you put it, "This file generates figures for tnse documentation"

File: docs/api/text/tsne.rst [hunk 7]
```diff
@@ -34,29 +34,18 @@ Now that the corpus is vectorized we can visualize it, showing the distribution
...
     # Don't color points with their classes
-    tsne = TSNEVisualizer()
+    tsne = TSNEVisualizer(labels=["documents"])
     tsne.fit(docs)
     tsne.poof()
 
 .. image:: images/tsne_no_labels.png
 
-Finally, we don't have to use class labels at all, instead we can use cluster membership from K-Means to label each document:
+This means we don't have to use class labels at all, instead we can use cluster membership from K-Means to label each document, looking for clusters of related text by their contents:
```
Comment (lwgray) [inline docs/api/text/tsne.rst]: This is just a stylistic comment.  It doesn't need to be change.  However, there is a lot of information in this sentence; try breaking it up into smaller chunks

File: yellowbrick/text/tsne.py [hunk 8]
```diff
@@ -252,13 +267,18 @@ def fit(self, X, y=None, **kwargs):
...
+        if y is not None:
+            self.classes_ = np.unique(y)
+        else:
```
Comment (lwgray) [inline yellowbrick/text/tsne.py]: elif statement ???

File: tests/test_text/test_tsne.py [hunk 9]
```diff
@@ -17,26 +17,32 @@
...
+from sklearn.datasets import make_classification
 from sklearn.feature_extraction.text import TfidfVectorizer
 
 
 ##########################################################################
 ## TSNE Tests
 ##########################################################################
 
-class TSNETests(unittest.TestCase, DatasetMixin):
+class TestTSNE(VisualTestCase, DatasetMixin):
```
Comment (ndanielsen) [inline tests/test_text/test_tsne.py]: Ben these are great tests!

File: tests/test_text/test_tsne.py [hunk 10]
```diff
@@ -45,26 +51,98 @@ def test_make_pipeline(self):
...
         labels = corpus.target
 
-        tsne = TSNEVisualizer()
+        tsne = TSNEVisualizer(random_state=8392, colormap='Set1')
         tsne.fit_transform(docs, labels)
+        
+        tol = 40 if six.PY3 else 55
+        self.assert_images_similar(tsne, tol=tol)
+
+    def test_make_classification_tsne(self):
```
Comment (ndanielsen) [inline tests/test_text/test_tsne.py]: perfect, this is exactly the bug that was reported!

File: tests/test_text/test_tsne.py [hunk 11]
```diff
@@ -45,26 +51,98 @@ def test_make_pipeline(self):
...
+        with pytest.raises(YellowbrickValueError):
+            tsne.fit(X,y)
+
+        ## more labels than classes
+        tsne = TSNEVisualizer(random_state=87, labels=['a', 'b', 'c', 'd'])
+        with pytest.raises(YellowbrickValueError):
+            tsne.fit(X,y)
+
+
+    def test_no_target_tsne(self):
```
Comment (ndanielsen) [inline tests/test_text/test_tsne.py]: great job on testing for these edge cases!

File: tests/test_text/test_tsne.py [hunk 12]
```diff
@@ -45,26 +51,98 @@ def test_make_pipeline(self):
...
+        ## produce random data
+        X, y = make_classification(n_samples=200, n_features=100,
+                               n_informative=20, n_redundant=10,
+                               n_classes=3, random_state=6897)
+
+        ## visualize data with t-SNE
+        tsne = TSNEVisualizer(random_state=64)
+        tsne.fit(X)
+
+        self.assert_images_similar(tsne, tol=0.1)
```
Comment (ndanielsen) [inline tests/test_text/test_tsne.py]: Any chance that we can add a test for handling of pandas dataframes? We promise support in the docs for them, and probably have tests to support it's use.

File: yellowbrick/text/tsne.py [hunk 13]
```diff
@@ -274,34 +294,45 @@ def draw(self, points, target=None, **kwargs):
...
         """
-        # Create the color mapping for the classes.
-        # TODO: Allow both colormap, listed colors, and palette definition
-        # See the FeatureVisualizer for more on this.
-        color_values = get_color_cycle()
-        classes = self.classes_ or [None]
-        colors = dict(zip(classes, color_values))
+        # Resolve the labels with the classes
+        labels = self.labels if self.labels is not None else self.classes_
+        if len(labels) != len(self.classes_):
```
Comment (wagner2010) [inline yellowbrick/text/tsne.py]: This is clear to me.  I understand it.

File: tests/test_text/test_tsne.py (hunk 12 above)
Comment (bbengfort) [reply tests/test_text/test_tsne.py]: Will do

File: yellowbrick/text/tsne.py (hunk 5 above)
Comment (lwgray) [reply yellowbrick/text/tsne.py]: I am not sure how to handle or even advise you on this because normally without a label, the legend wouldn't be drawn even if you specify ax.legend(...)
